### ISSUE TITLE
Making Polars Work With Other Frameworks

### ISSUE THREAD

**Issue by Killpit** [pola-rs/polars#7703]
### Problem description

I really like polars, but when I have to work with numpy, matplotlib, plotly, scikit-learn, PyTorch, TensorFlow and other data science and visualization libraries, it simple doesn't work. I'd recommend adding features so that it supports these libraries and frameworks.

**Comment by ghuls** [comment]
Polars supports the DataFrame interchange protocol via PyArrrow interchange implementation,
So all downstream libraries should just call `df.__dataframe()` instead of implementing logic for every dataframe library itself.

https://data-apis.org/dataframe-protocol/latest/purpose_and_scope.html

```python
import polars as pl
import pandas as pd
import pyarrow as pa
import pyarrow.interchange

In [54]: df
Out[54]: 
shape: (2, 1)
┌─────┐
│ x   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
│ 2   │
└─────┘

In [55]: pl.from_dataframe(df.__dataframe__())
Out[55]: 
shape: (2, 1)
┌─────┐
│ x   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
│ 2   │
└─────┘

In [56]: pd.api.interchange.from_dataframe(df.__dataframe__())
Out[56]: 
   x
0  1
1  2

In [61]: pa.interchange.from_dataframe(df.__dataframe__())
Out[61]: 
pyarrow.Table
x: int64
----
x: [[1,2]]

In [62]: pa.interchange.from_dataframe(df.__dataframe__()).__dataframe__()
Out[62]: <pyarrow.interchange.dataframe._PyArrowDataFrame at 0x7f0fa8712a00>

In [63]: pd.api.interchange.from_dataframe(df.__dataframe__())
Out[63]: 
   x
0  1
1  2

In [64]: pd.api.interchange.from_dataframe(df.__dataframe__()).__dataframe__()
Out[64]: <pandas.core.interchange.dataframe.PandasDataFrameXchg at 0x7f104c7c5340>

In [65]: df
Out[65]: 
shape: (2, 1)
┌─────┐
│ x   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
│ 2   │
└─────┘


# Create a dataframe via dataframe interchange protocol called on a pandas dataframe.
In [66]: pl.from_dataframe(pd.api.interchange.from_dataframe(df.__dataframe__()).__dataframe__())
Out[66]: 
shape: (2, 1)
┌─────┐
│ x   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
│ 2   │
└─────┘

# Create a dataframe via dataframe interchange protocol called on a pyarrow Table.
In [67]: pl.from_dataframe(pa.interchange.from_dataframe(df.__dataframe__()).__dataframe__())
Out[67]: 
shape: (2, 1)
┌─────┐
│ x   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
│ 2   │
└─────┘

```

**Comment by Killpit** [comment]
But you can't use polars with other libraries, especially with visualisation libraries

**Comment by lucazanna** [comment]
Hi @Killpit , 

Here is my 2 cents : as Polars supports the dataframe interchange protocol, it’s up to those libraries to start supporting the protocol.

As soon as they do, they will become compatible with Polars and with all other dataframe libraries supporting the protocol. So it’s a win for them 

In the meanwhile, you can convert to Polars to Pandas as Pandas is the most widely supported dataframe library.

the conversion is quite straightforward with .to_pandas()

You can also convert to Pandas with an arrow backend with .to_pandas(with_pyarrow_extension_array = True)
this should give you fast conversion times, so the conversion will not slow down you code

**Comment by datapythonista** [comment]
I think it'd be good to consider if besides making Polars work with the ecosystem via the interchange protocol, it makes sense to allow an API for certain functionality like plotting. Not to include plotting in Polars itself (which I don't think it's a good idea), but to allow users to have a nicer syntax in a nice and standard way.

For example, consider these two examples:

### Example 1

```python
(polars.scan_csv('transactions.csv', null_values='null', try_parse_dates=True)
       .filter(_('trans_group_en') == 'Sales')
       .filter(_('property_sub_type_en') == 'Flat')
       .with_columns((_('actual_worth') / _('procedure_area')).alias('price_sqm_aed'))
       .sort('instance_date')
       .groupby_dynamic('instance_date', every='1mo')
       .agg(_('price_sqm_aed').mean())
       .plot(x='instance_date', y='price_sqm_aed', range_y=(0, 15_000)))
```

### Example 2

```python
import plotly.express

plotly.express.line(polars.scan_csv('transactions.csv', null_values='null', try_parse_dates=True)
                          .filter(_('trans_group_en') == 'Sales')
                          .filter(_('property_sub_type_en') == 'Flat')
                          .with_columns((_('actual_worth') / _('procedure_area')).alias('price_sqm_aed'))
                          .sort('instance_date')
                          .groupby_dynamic('instance_date', every='1mo')
                          .agg(_('price_sqm_aed').mean()))
                    x='instance_date',
                    y='price_sqm_aed',
                    range_y=(0, 15_000))
```

I think for users option 1 it's much nicer. Also if the dataframe is assigned to a variable I think there is still an advantage of using `df.plot(...)`  instead of `px.line(df, ...)`. This can easily be implemented by users via:

```python
import plotly.express

def plot(self, x, y, kind='line', **kwargs):
    return getattr(plotly.express, kind)(data_frame=self.collect().to_pandas(),
                                         x=x, y=y,
                                         **kwargs)

polars.LazyFrame.plot = plot
```

But I don't think regular users should be encourage to inject methods themselves, as it opens a can of worms and allows for Polars code that is not really Polars. I think a better solution is if Polars provides a standard "empty" `plot` function and a way for third party plugins to be able to connect to it and provide the plotting functionality for the ecosystem of plotting libraries. Python entrypoints or `polars.Config` could be used to connect Polars with the plugins.

**Comment by stinodego** [comment]
I'm closing this as there is no clear action to take here. Feel free to open a feature request for the plotting functionality.

**Comment by ivanstepanovftw** [comment]
Dataframe is a data type and it is not able to plot itself. As your cat cannot do homework for you. Unless you told `cat.do_my_homework()`...
